File: PyTutorials/operators.py
'''
There's several operators in python, here's a list of em:

First there's the Arithmetic operators:

        Operator  |  Example
                  |
            +     |     1 + 2        list1 + list2
       (Addition) |   var1 + var2    string1 + string2 <---- this one's known as concatenation
                  |
                  |
            -     |     1 - 2        <no '-' operator for lists or strings>
     (Subtraction)|   var1 - var2
                  |
                  |
            *     |     1 * 2        string * n <--------- repeats the string n-number of times,  for example:
  (Multiplication)|   var1 * var2    "hello"*5 = "hellohellohellohellohello" (we can do the same with lists as well...)
                  |
                  |
            /     |     1 / 2       <no '/' operator for lists or strings>
       (Division) |   var1 / var2
                  |
                  |
            %     |     1 % 2       <no '%' operator for lists or strings>
       (Modulus)  |   var1 % var2
                  |
                  |
            //    |     1 // 2      <no '//' operator for lists or strings>
  (Floor-Division)|   var1 // var2
                  |
                  |
            **    |     1 ** 2      <no '**' operator for lists or strings>
  (Exponentiation)|   var1 ** var2

The (Integer-Division) Floor-Division operator truncates the decimal part of the result of division of numbers.

The Modulus operator gives us the remainder of 2 numbers when they're divided.
'''

# Examples of useage of Arithmetic operators:

print("Adding numbers:       1 + 2 = ", 1 + 2)
print("Adding(concatenating) strings: 'string1' + 'string2' = ", "string1" + "string2" )
print("Adding lists:         [1, 2, 3] + [4, 5, 6] = ", [1, 2, 3] + [4, 5, 6])

print("\nSubtracting numbers: 1 - 2 = ", 1 - 2)

print("\nMultipying numbers: 123 * 2784 = ", 123 * 2784)
print("Repeating strings: 'string1' * 5 ", "string1" * 5)
print(" '*' With lists: [0] * 10 = ", [0] * 10)

print("\nDividing:\nSimple-Division: 5/2 = ", 5/2, "\nFloor-Division: 5//2 = ", 5//2)
print("Modulus: 124%3 = ", 124%3)

print("\nExponentiation: 5**2 = ", 5**2)

'''
Next up, there's the Logical and Comparison operators:

        Operator  |  Example
                  |
            >     |     1 > 2        list1 > list2 <--------| In both lists and strings alike, the <(=) and >(=) operators
                  |   var1 > var2    string1 > string2 <----| will compare the elements/characters one-by-one.
                  |                                         | The first element to give false will cause the operation
            <     |     1 - 2        list1 < list2 <--------| to give the output/result as False. If no such occurence
                  |   var1 - var2    string1 < string2 <----| is there, the output/result is True.
                  |
            >=    |     1 >= 2       list1 >= list2 <-------| The main difference between the < and <=, and the > and >=
                  |   var1 >= var2   string1 >= string2 <---| operators is the strict inequality of < and >, i.e.
                  |                                         | 1 < 1 == False but 1 <= 1 == True.
            <=    |     1 <= 2       list1 <= list2 <-------|
                  |   var1 <= var2   string1 <= string2 <---|
                  |
            !=    |     1 != 2       list1 != list2 <-------| The != operator checks if the 2 variables/strings/lists/anything
                  |   var1 != var2   string1 != string2 <---| are equal and returns the opposite i.e. 1!=2 == True.
                  |
            ==    |     1 == 2       list1 == list2 <-------| The == ( equivalance ) operator helps check if something on the
                  |   var1 == var2   string1 == string2 <---| left side's equal to something on the right side.
                  |
            =     |   var = value   <-----------------------| This is the assignment operator, and is used to assign values to
                  |                                         | variables.
                  |
            and   | bool_1 and bool_2 ... <-----------------| Results in a 'True' boolean value if ALL of the operands are 'True'.
                  |
                  |
            or    | bool_1 or bool_2 ... <------------------| Results in a 'True' boolean value if any of the operands are 'True'.
                  |
                  |
            not   | not bool ... <--------------------------| Results in a 'True' value if 'bool' if 'False' and vise-versa.
'''

print("1 > 2 = ", 1 > 2)
print("1 < 2 = ", 1 < 2)

print("\n1 <= 2 = ", 1 <= 2)
print("1 >= 2 = ", 1 >= 2)
print("1 <= 1 = ", 1 <= 2)

print("\n1 != 2 = ", 1 != 2)
print("1 == 2 = ", 1 == 2)
print("1 == 1 = ", 1 == 1)

print("\nTrue and True = ", True and True)
print("True and False = ", True and False)
print("True and True and True ... = ", True and True and True)

print("\nTrue or True = ", True or True)
print("True or False = ", True or False)
print("True or False or True ... = ", True or False or True)

print("\nnot True = ", not True)
print("not False = ", not False)
print("\nTrue and not False = ", True and not False)


'''
Next up, there's the Bitwise operators:

        Operator        |  Example
                        |
            &           |   0110 & 0101 <---------------------| Each bit goes through an AND operation.
     (Bitwise - AND)    |    => 0100
                        |
                        |
            |           |   1010 | 0101 <---------------------| Each bit goes through an OR operation.
     (Bitwise - OR)     |    => 1111
                        |
                        |
            ~           |      ~1001 <------------------------| Each bit goes through an NOT operation.
     (Bitwise - NOT)    |    => 0110
                        |
                        |
            ^           |  1001 ^ 0101 <----------------------| Each bit goes through an XOR operation.
     (Bitwise - XOR)    |    => 1100
                        |
                        |
            >>          |  0101 >> n <------------------------| Performs a left shift on the number upto n bits.
   (Bitwise Left Shift) |
                        |
                        |
            <<          |  0101 << n <------------------------| Performs a right shift on the number upto n bits.
  (Bitwise Right Shift) |
'''

# There's some special operators as well!
# They are:
# 1) is <---------- True if the operands are identical (refer to the same object)
# 2) in <---------- True if value/variable is found in the sequence


# NOTE: Here's an operator precedence chart for reference: https://docs.python.org/3/reference/expressions.html#operator-precedence

File: CTutorials/operators.c
#include<stdio.h>

/*
There's several operators in python, here's a list of em:

First there's the Arithmetic operators:

        Operator  |  Example
                  |
            +     |     1 + 2
       (Addition) |   var1 + var2
                  |
                  |
            -     |     1 - 2
     (Subtraction)|   var1 - var2
                  |
                  |
            *     |     1 * 2
  (Multiplication)|   var1 * var2
                  |
                  |
            /     |     1 / 2
       (Division) |   var1 / var2
                  |
                  |
            %     |     1 % 2
       (Modulus)  |   var1 % var2
                  |
                  |
            ++    |     ++var <----------------------| ++var increments(by one) before a value is returned
      (Increment) |     var++ <----------------------| var++ increments(by one) after a value is returned
                  |
                  |
            --    |   --var  <-----------------------| --var decrements(by one) before a value is returned
      (Decrement) |   var--  <-----------------------| var-- decrements(by one) after a value is returned


The Modulus operator gives us the remainder of 2 numbers when they're divided.
*/



/*
Next up, there's the Logical and Comparison operators:

        Operator  |  Example
                  |
            >     |     1 > 2
                  |   var1 > var2
                  |
            <     |     1 - 2
                  |   var1 - var2
                  |
            >=    |     1 >= 2
                  |   var1 >= var2
                  |
            <=    |     1 <= 2
                  |   var1 <= var2
                  |
            !=    |     1 != 2
                  |   var1 != var2
                  |
            ==    |     1 == 2
                  |   var1 == var2
                  |
            &&    |   bool_1 && bool_2 <-----------------| performs a logical AND operation on the booleas.
                  |
                  |
            ||    |   bool_1 || bool_2 <-----------------| performs a logical OR operation on the booleans.
                  |
                  |
            !     |       !bool <------------------------| performs a logical NOT operation.
*/


/*
Next up, there's the Bitwise operators:

        Operator        |  Example
                        |
            &           |   0110 & 0101 <---------------------| Each bit goes through an AND operation.
     (Bitwise - AND)    |    => 0100
                        |
                        |
            |           |   1010 | 0101 <---------------------| Each bit goes through an OR operation.
     (Bitwise - OR)     |    => 1111
                        |
                        |
            ~           |      ~1001 <------------------------| Each bit goes through an NOT operation.
     (Bitwise - NOT)    |    => 0110
                        |
                        |
            ^           |  1001 ^ 0101 <----------------------| Each bit goes through an XOR operation.
     (Bitwise - XOR)    |    => 1100
                        |
                        |
            >>          |  0101 >> n <------------------------| Performs a left shift on the number upto n bits.
   (Bitwise Left Shift) |
                        |
                        |
            <<          |  0101 << n <------------------------| Performs a right shift on the number upto n bits.
  (Bitwise Right Shift) |
*/

int main() {
    // Examples of useage of Arithmetic operators:

    printf("Adding numbers:       1 + 2 = %d\n", 1 + 2);

    printf("\nSubtracting numbers: 1 - 2 = %d\n", 1 - 2);

    printf("\nMultipying numbers: 123 * 2784 = %d\n", 123 * 2784);

    printf("\nModulus: 124%%3 = %d\n", 124%3);

    int var = 10; // just a variable to for illustrating increment/decrement.
    printf("\nvar++: %d\n", var++);
    var = 10; // resetting the value of "var"
    printf("++var: %d\n", ++var);
    var = 10; // resetting the value of "var"

    printf("\nvar--: %d\n", var--);
    var = 10; // resetting the value of "var"
    printf("--var: %d\n", --var);
    var = 10; // resetting the value of "var"


    printf("\n1 > 2 = %d\n", 1 > 2);
    printf("1 < 2 = %d\n", 1 < 2);

    printf("\n1 <= 2 = %d\n", 1 <= 2);
    printf("1 >= 2 = %d\n", 1 >= 2);
    printf("1 <= 1 = %d\n", 1 <= 2);

    printf("\n1 != 2 = %d\n", 1 != 2);
    printf("1 == 2 = %d\n", 1 == 2);
    printf("1 == 1 = %d\n", 1 == 1);

    printf("\nTrue(1) && True(1) = %d\n", 1 && 1);
    printf("True(1) && False = %d\n", 1 && 0);
    printf("True(1) && True(1) && True(1) ... = %d\n", 1 && 1 && 1);
    printf("True(1) && False(0) && True(1) ... = %d\n", 1 && 0 && 1);

    printf("\nTrue(1) || True(1) = %d\n", 1 || 1);
    printf("True(1) || False(0) = %d\n", 1 || 0);
    printf("True(1) || True(1) || True(1) ... = %d\n", 1 || 1 || 1);
    printf("True(1) || False(0) || True(1) ... = %d\n", 1 || 0 || 1);

    printf("\n !True(1) = %d\n", !1);
    printf("!False(0) = %d\n", !0);
    printf("True(1) && !0 = %d\n",1 && !0);

}
// NOTE: Here's an operator precedence chart for reference: https://docs.microsoft.com/en-us/cpp/c-language/precedence-and-order-of-evaluation?view=msvc-160#precedence-and-associativity-of-c-operators
